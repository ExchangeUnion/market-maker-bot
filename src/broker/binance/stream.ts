import { Stream } from '../stream';
import WebSocket from 'ws';
import { Logger } from '../../logger';

// If we haven't received a ping during this time we'll consider
// socket dead.
const HEARTBEAT_TIMEOUT = 60 * 1000 * 3 + 1000;
// Since Binance allows to keep a single socket alive for maximum of 24h we're
// recrating the socket after the specified amount of time
const RESTART_TIMEOUT = 60 * 60 * 1000;

class BinanceStream extends Stream {
  private isAlive = false;
  private socket!: WebSocket;
  private pingTimeout!: ReturnType<typeof setTimeout>;
  private restartTimeout!: ReturnType<typeof setTimeout>;

  constructor(private logger: Logger, private tradingPair: string) {
    super();
  }

  public start = (): Promise<void> => {
    return new Promise((resolve) => {
      clearTimeout(this.pingTimeout);
      clearTimeout(this.restartTimeout);
      const url = `wss://stream.binance.com:9443/ws/${this.tradingPair.toLowerCase()}@aggTrade`;
      this.socket = new WebSocket(url);

      this.socket.onopen = () => {
        this.logger.info(`${this.tradingPair} established connection to ${url}`);
      };

      this.socket.on('ping', this.heartbeat);
      this.socket.on('open', this.heartbeat);
      this.socket.on('error', (e) => {
        this.logger.error(`error from the socket ${e}`);
      });

      this.socket.onclose = this.onClose;
      this.socket.onmessage = this.onMessage;
      this.restartTimeout = setTimeout(async () => {
        this.logger.warn(`${this.tradingPair} socket timeout reached - restarting`);
        this.socket.terminate();
        await this.restart();
      }, RESTART_TIMEOUT);
      resolve();
    });
  }

  public close = (): Promise<void> => {
    clearTimeout(this.restartTimeout);
    this.socket.terminate();
    return new Promise((resolve) => {
      const isAliveCheck = setInterval(() => {
        if (!this.isAlive) {
          clearInterval(isAliveCheck);
          resolve();
        }
      });
    });
  }

  private onMessage = (event: WebSocket.MessageEvent) => {
    const aggTrade = JSON.parse(event.data.toString());
    const { p: priceString } = aggTrade;
    const price = parseFloat(priceString);
    this.emit('price', this.tradingPair, price);
  }

  private onClose = (event: WebSocket.CloseEvent) => {
    clearTimeout(this.pingTimeout);
    this.isAlive = false;
    if (event.reason) {
      this.logger.info(`${this.tradingPair} stream closed with reason: ${event.reason}`);
    } else {
      this.logger.info(`${this.tradingPair} stream closed`);
    }
  }

  private restart = async () => {
    try {
      // restart if heartbeat fails
      await this.start();
    } catch (e) {
      this.logger.error(`${this.tradingPair} failed to restart stream: ${e}`);
    }
  }

  private heartbeat = () => {
    this.logger.info(`heartbeat from ${this.tradingPair} socket - isAlive: ${this.isAlive}`);
    this.isAlive = true;
    clearTimeout(this.pingTimeout);
    this.pingTimeout = setTimeout(async () => {
      this.isAlive = false;
      this.socket.terminate();
      await this.restart();
    }, HEARTBEAT_TIMEOUT);
  }

}

export { BinanceStream };
