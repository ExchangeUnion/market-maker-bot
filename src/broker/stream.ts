import { EventEmitter } from 'events';

interface Stream {
  on(event: 'price', listener: (tradingPair: string, price: number) => void): this;
  emit(event: 'price', tradingPair: string, price: number): boolean;
}

abstract class Stream extends EventEmitter {
  public abstract start(): Promise<void>;
  public abstract close(): Promise<void>;
}

export { Stream };
