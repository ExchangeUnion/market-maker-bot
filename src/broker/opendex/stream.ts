import { Stream } from '../stream';

class OpenDexStream extends Stream {
  public start = () => {
    return Promise.resolve();
  }
  public close = () => {
    return Promise.resolve();
  }
}

export { OpenDexStream };
