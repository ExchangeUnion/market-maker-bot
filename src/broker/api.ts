import { OrderSide, OrderType, OrderStatus } from '../enums';
import {
  MarketOrderRequest,
  LimitOrderRequest,
  StopLimitOrderRequest,
} from './exchange';
import { BinanceOrder } from './binance/order';
import { OpenDexOrder } from './opendex/order';

export type BalanceProperties = {
  free: number;
  locked: number;
  maxsell?: number;
  maxbuy?: number;
};

export type Balance = BalanceProperties & {
  asset: string;
};

export type OrderRequest = {
  baseAsset: string;
  quoteAsset: string;
  side: OrderSide;
  type: OrderType;
  quantity: number;
  price?: number;
  stopPrice?: number;
};

export type QueryOrderResponse = {
  tradingPair: string;
  status: OrderStatus;
};

export type QueryOrderRequest = {
  tradingPair: string;
  orderId: string;
};

export type TradingLimits = {
  asset: string;
  maxsell: number;
  maxbuy: number;
};

export type CancelOrderRequest = QueryOrderRequest;

abstract class ExchangeAPI {
  public abstract start(): Promise<void>;
  public abstract stop(): Promise<void>;
  public abstract getAssets(): Promise<Balance[]>;
  public abstract newOrder(
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest
  ): BinanceOrder | OpenDexOrder;
  public abstract queryOrder(
    queryOrderRequest: QueryOrderRequest
  ): Promise<QueryOrderResponse>;
  public abstract cancelOrder(
    cancelOrderRequest: CancelOrderRequest
  ): Promise<boolean>;
}

export { ExchangeAPI };
