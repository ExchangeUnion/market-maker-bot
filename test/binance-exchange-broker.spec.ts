import { ExchangeBroker } from '../src/broker/exchange';
import {
  ExchangeType,
  OrderSide,
  OrderType,
} from '../src/enums';
import { BinanceStream } from '../src/broker/binance/stream';
import { BinanceAPI } from '../src/broker/binance/api';
import { Level, Logger } from '../src/logger';
import { mockAccountInfo } from '../mock-data/account-info';

const loggers = Logger.createLoggers(Level.Warn);

jest.mock('../src/broker/binance/api');
const binanceAPIstart = BinanceAPI.prototype.start = jest.fn();
BinanceAPI.prototype.stop = jest.fn();
BinanceAPI.prototype.cancelOrder = jest.fn();
BinanceAPI.prototype.newOrder = jest.fn();

jest.mock('../src/broker/binance/stream');

describe('Broker', () => {

  describe('Binance', () => {

    describe('verify configuration', () => {

      test('throws without api key and secret', async () => {
        expect.assertions(1);
        try {
          const broker = new ExchangeBroker({
            exchange: ExchangeType.Binance,
            logger: loggers.binance,
          });
          await broker.start();
        } catch (e) {
          expect(e).toMatchSnapshot();
        }
      });

      test('throws without api key', async () => {
        expect.assertions(1);
        try {
          const broker = new ExchangeBroker({
            exchange: ExchangeType.Binance,
            apiSecret: 'abc',
            logger: loggers.binance,
          });
          await broker.start();
        } catch (e) {
          expect(e).toMatchSnapshot();
        }
      });

      test('throws without api secret', async () => {
        expect.assertions(1);
        try {
          const broker = new ExchangeBroker({
            exchange: ExchangeType.Binance,
            apiKey: '123',
            logger: loggers.binance,
          });
          await broker.start();
        } catch (e) {
          expect(e).toMatchSnapshot();
        }
      });

      test('starts with api key and secret specified', async () => {
        const broker = new ExchangeBroker({
          exchange: ExchangeType.Binance,
          apiKey: '123',
          apiSecret: 'abc',
          logger: loggers.binance,
        });
        await broker.start();
        expect(binanceAPIstart).toHaveBeenCalledTimes(1);
      });

    });

    describe('getAssets', () => {

      it('returns asset allocation', async () => {
        const broker = new ExchangeBroker({
          exchange: ExchangeType.Binance,
          apiKey: '123',
          apiSecret: 'abc',
          logger: loggers.binance,
        });
        const getAssetsMock = BinanceAPI.prototype.getAssets = jest
          .fn()
          .mockReturnValue(mockAccountInfo.data.balances);
        await broker.start();
        const ownedAssets = await broker.getAssets();
        expect(getAssetsMock).toHaveBeenCalledTimes(1);
        expect(ownedAssets).toMatchSnapshot();
      });

    });

    describe('newOrder', () => {

      test('it creates a new stop-limit order', async () => {
        const broker = new ExchangeBroker({
          exchange: ExchangeType.Binance,
          apiKey: '123',
          apiSecret: 'abc',
          logger: loggers.binance,
        });
        await broker.start();
        const orderRequest = {
          orderId: '123',
          baseAsset: 'ETH',
          quoteAsset: 'BTC',
          orderType: OrderType.StopLimit,
          orderSide: OrderSide.Buy,
          quantity: 0.005,
          price: 0.023006,
          stopPrice: 0.025410,
        };
        await broker.newOrder(orderRequest);
        expect(BinanceAPI.prototype.newOrder).toHaveBeenCalledTimes(1);
      });

    });

    describe('getPrice', () => {
      let broker: ExchangeBroker;
      const onBTCUSDTpriceChange = jest.fn();
      const startStream = BinanceStream.prototype.start = jest.fn();
      const onListener = BinanceStream.prototype.on = jest.fn();
      const closeListener = BinanceStream.prototype.close = jest.fn();

      beforeEach(async () => {
        broker = new ExchangeBroker({
          exchange: ExchangeType.Binance,
          apiKey: '123',
          apiSecret: 'abc',
          logger: loggers.binance,
        });
        await broker.start();
        jest.resetAllMocks();
      });

      test('getPrice creates price stream for tradingPair', async () => {
        expect(broker['priceStreams'].size).toEqual(0);
        const tradingPair = 'BTCUSDT';
        await broker.getPrice(tradingPair, onBTCUSDTpriceChange);
        expect(broker['priceStreams'].size).toEqual(1);
        expect(broker['priceStreams'].has(tradingPair)).toBeTruthy();
      });

      test('getPrice does not create duplicate streams', async () => {
        const tradingPair = 'BTCUSDT';
        await broker.getPrice(tradingPair, onBTCUSDTpriceChange);
        await broker.getPrice(tradingPair, onBTCUSDTpriceChange);
        expect(BinanceStream).toHaveBeenCalledTimes(1);
        expect(startStream).toHaveBeenCalledTimes(1);
        expect(onListener).toHaveBeenCalledTimes(2);
      });

      describe('close', () => {

        it('destroys all streams and closes API', async () => {
          expect(broker['priceStreams'].size).toEqual(0);
          await broker.getPrice('BTCUSDT', () => {});
          await broker.getPrice('ETHBTC', () => {});
          expect(broker['priceStreams'].size).toEqual(2);
          // close destroys pending streams
          const apiClose = BinanceAPI.prototype.stop = jest.fn();
          await broker.close();
          expect(closeListener).toHaveBeenCalledTimes(2);
          expect(apiClose).toHaveBeenCalledTimes(1);
        });

      });

    });

  });

});
