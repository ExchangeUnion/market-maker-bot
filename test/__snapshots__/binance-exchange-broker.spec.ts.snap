// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Broker Binance getAssets returns asset allocation 1`] = `
Array [
  Object {
    "asset": "BTC",
    "free": "0.10054433",
    "locked": "0.00000000",
  },
  Object {
    "asset": "LTC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ETH",
    "free": "0.00087248",
    "locked": "0.00000000",
  },
  Object {
    "asset": "NEO",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BNB",
    "free": "0.00858866",
    "locked": "0.00000000",
  },
  Object {
    "asset": "QTUM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "EOS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "SNT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BNT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "GAS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BCC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "USDT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "HSR",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "OAX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DNT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "MCO",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ICN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ZRX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "OMG",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "WTC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "YOYO",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "LRC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "TRX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "SNGLS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "STRAT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BQX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "FUN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "KNC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "CDT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "XVG",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "IOTA",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "SNM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "LINK",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "CVC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "TNT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "REP",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "MDA",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "MTL",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "SALT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "NULS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "SUB",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "MTH",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ADX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ETC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ENG",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ZEC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "AST",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "GNT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DGD",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BAT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DASH",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "POWR",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BTG",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "REQ",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "XMR",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "EVX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "VIB",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ENJ",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "VEN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ARK",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "XRP",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "MOD",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "STORJ",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "KMD",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "RCN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "EDO",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DATA",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DLT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "MANA",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "PPT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "RDN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "GXS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "AMB",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ARN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BCPT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "CND",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "GVT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "POE",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BTS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "FUEL",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "XZC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "QSP",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "LSK",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BCD",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "TNB",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ADA",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "LEND",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "XLM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "CMT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "WAVES",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "WABI",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "GTO",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ICX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "OST",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ELF",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "AION",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "WINGS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BRD",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "NEBL",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "NAV",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "VIBE",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "LUN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "TRIG",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "APPC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "CHAT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "RLC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "INS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "PIVX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "IOST",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "STEEM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "NANO",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "AE",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "VIA",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BLZ",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "SYS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "RPX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "NCASH",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "POA",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ONT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ZIL",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "STORM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "XEM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "WAN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "WPR",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "QLC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "GRS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "CLOAK",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "LOOM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BCN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "TUSD",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ZEN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "SKY",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "THETA",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "IOTX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "QKC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "AGI",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "NXS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "SC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "NPXS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "KEY",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "NAS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "MFT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DENT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ARDR",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "HOT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "VET",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DOCK",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "POLY",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ONG",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "PHX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "HC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "GO",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "PAX",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "RVN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DCR",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "USDC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "MITH",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BCHABC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BCHSV",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "REN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BTT",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "USDS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "FET",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "TFUEL",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "CELR",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "MATIC",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ATOM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "PHB",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ONE",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "FTM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BTCB",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "USDSB",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "CHZ",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "COS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ALGO",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ERD",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DOGE",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BGBP",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "DUSK",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "ANKR",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "WIN",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "TUSDB",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "COCOS",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "PERL",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "TOMO",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BUSD",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BAND",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "BEAM",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "HBAR",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
  Object {
    "asset": "XTZ",
    "free": "0.00000000",
    "locked": "0.00000000",
  },
]
`;

exports[`Broker Binance verify configuration throws without api key 1`] = `[Error: Binance exchange type requires you to specify the apiKey and apiSecret]`;

exports[`Broker Binance verify configuration throws without api key and secret 1`] = `[Error: Binance exchange type requires you to specify the apiKey and apiSecret]`;

exports[`Broker Binance verify configuration throws without api secret 1`] = `[Error: Binance exchange type requires you to specify the apiKey and apiSecret]`;
